Disk Shader Caching

jroweboy opened the pull request:
Based on the excellent work from Rodrigo, this brings over much of the framework that he wrote but with obvious changes to support 3DS shaders instead.

Design:
There are two caches with different reasons to wipe the cache. The transferrable cache is meant to be device agnostic, and store everything from the PICA state needed to generate the shaders. For FS and VS, we store the entire PICA registers, as its simpler to handle storing the same configuration for both types. Additionally, for VS, we store the shader program and shader swizzle data from memory as well. As this cache should be essentially unchanged from version to version, it is only invalidated when either 1) a manual version bump occurs or 2) the cache is corrupted and a shader fails to build

The precompiled cache stores both generated GLSL and the resulting binary cache. This cache is only built during the game launch when the transferrable cache is loaded. If this cache is present, we will read all of the elements from the transferrable cache, and attempt to match up and "inject" the precompiled shaders into the in-memory ShaderCaches. If this cache is present, it should speed up loading as we don't need to run a full shader compile for all of the shaders. This cache is invalidated in the following circumstances: 1) Precompiled shader binary is rejected by the driver 2) any of the long list of files related to shader generation is changed (causing the generated version hash in cmake to not match) 3) the cache is missing both the decompiled glsl and the precompiled binary.

Currently there is *NOT* a loading screen as I want people to test it out first and focus on the actual implementation. I would rather avoid a loading screen entirely and instead do async compilation on a separate context, but if we can't work that out in time before this gets merged, I can tack on the loading screen I made for yuzu.

UI Screenshot
![citra-qt_2019-09-09_19-51-34](https://user-images.githubusercontent.com/952515/64578028-5f7bf080-d33b-11e9-8c19-609c319882c8.png)


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/4923)
<!-- Reviewable:end -->

chris062689 commented:
Are there any changes to settings windows?
Any option to enable / disable disk shader caching?

wwylele commented:
Posting this message from discord here in case we forget: the introduction of vfs is  unnecessary. Can implement a simple vector appender (or not at all) for this use case.

wwylele commented on src/video_core/renderer_opengl/gl_shader_manager.cpp:
```diff
@@ -3,13 +3,81 @@
+
+    auto shader = OGLProgram();
+    shader.handle = glCreateProgram();
+    glProgramParameteri(shader.handle, GL_PROGRAM_SEPARABLE, GL_TRUE);
```
This is an extension. Did you make this optional? (Sorry if I missed some code)

jroweboy replied:
The disk cache only loads with separable shader objects enabled. I didn't wanna play around with caching full shader programs at this point in time ^^; Added a log in case they don't support it so theres some feedback. FYI I'm intentionally still saving the shaders to the transferable cache as i don't see a reason not too.

wwylele commented on src/video_core/renderer_opengl/gl_shader_disk_cache.h:
```diff
@@ -0,0 +1,213 @@
+    /// Loads current game's precompiled cache. Invalidates on failure.
+    std::pair<ShaderDecompiledMap, ShaderDumpsMap> LoadPrecompiled();
+
+    /// Removes the transferable (and precompiled) cache file.
```
This comment indicates the function name is bad. Maybe rename it to InvalidateAll?

jroweboy replied:
Hmm, I'm not sure I agree, but I'll change it to save on argument. I think it logically follows that if the Transferable cache is invalidated, then you have to invalidate the Precompiled as well. Renamed it anyway 👍

jroweboy commented:
@chris062689 Added ui screenshot (Its a single checkbox in graphics)

@wwylele Addressed all concerns and fixed any known issues. Putting it on canary for user testing 👍

FearlessTobi commented:
Just a nit, but can't we at least keep the checkbox labeling the same as it is in yuzu?

BreadFish64 commented on src/video_core/renderer_opengl/gl_shader_manager.cpp:
```diff
@@ -257,45 +349,68 @@ class ShaderProgramManager::Impl {
+    // Save VS to the disk cache if its a new shader
+    if (result) {
+        auto& disk_cache = impl->disk_cache;
+        ProgramCode program_code{setup.program_code.begin(), setup.program_code.end()};
```
Isn't this equivalent to just `ProgramCode program_code{setup.program_code};`

BreadFish64 commented on src/video_core/renderer_opengl/gl_shader_manager.h:
```diff
@@ -6,14 +6,21 @@
 namespace OpenGL {
 
+class ShaderDiskCacheOpenGL;
+
```
Neither of these appear to used in this header.

jroweboy commented on src/common/CMakeLists.txt:
```diff
@@ -1,45 +1,50 @@
+      -P "${CMAKE_SOURCE_DIR}/CMakeModules/GenerateSCMRev.cmake"
+    DEPENDS
+      # WARNING! It was too much work to try and make a common location for this list,
+      # so if you need to change it, please update CMakeModules/GenerateSCMRev.cmake as well
```
Eh, a new developer won't be changing that file before changing src/common so i don't see a reason too.

jroweboy commented on src/common/file_util.cpp:
```diff
@@ -770,6 +769,103 @@ void SplitFilename83(const std::string& filename, std::array<char, 9>& short_nam
+               path.end());
+    return std::string(RemoveTrailingSlash(path));
+}
+
```
we support mac sadly, so i don't want to deal with that. last i checked, mac had it under a different header, and it wasn't complete? but i could be wrong.

jroweboy commented:
@weihuoya I'm aware, but any android specific changes aren't a part of this PR and will be published with the rest of the android changes.

John-Gee commented:
Hello,

I have a weird issue that I think is related to this PR, at least my bisection seems to say so as it does not happen on Canary with building at the previous commit. But disabling disk caching in the options changes nothing.

When I switch virtual desktops, Citra's performance drop to probably about the same as if I disable the hardware renderer, obviously I cannot see the FPS counter but the audio is super slow/broken. Now interestingly just alt-tabbing out of Citra does not cause that issue, only changing virtual desktops.

I've quickly tried with melonDS and The Dragon's Trap and I don't see that issue there, so I don't think it's my system's fault, but not completely sure either.

Specs wise:
I have a 4790k with a Radeon RX 580, running on Arch Linux 64b, using Mesa for the driver, and my DE is KDE's Plasma 5.17.4.

log: https://gist.github.com/John-Gee/9d38443f53575e1f19f9940d414b2d0d
The game I've tried in this log is Fire Emblem Echoes, but I've just tried with the much simpler (I think) Cave Story and it was the same.

I'll be happy to add any log/information needed or jump on chat if that's more helpful.

Thank you!

John-Gee commented:
Based on:
https://github.com/citra-emu/citra/pull/4940#issuecomment-560969476

If I understand it correctly, I'm guessing the issue is from the 4940 PR, not this one, but since they are merged, it's hard for me to try either way.

jroweboy commented:
@John-Gee I'm going to be merging #4940 since replicating the issue you mentioned is hard for me since i develop on windows and that PR is holding up a lot of upcoming changes. I'd still like to take a look at fixing that at some point, so please just open a new issue with your findings so we can track it down.

jroweboy commented:
Known issues:
- The cache doesn't take into account accurate mul setting, which can cause some weird situations with games that rely on it, but have a cache that wasn't compiled with it on. Currently investigating what the best route for fixing that is
- A small number of games have a large number of shaders so load times are exasperated on those. Tomorrow I plan on looking at a fix for that, and if it doesn't look likely, then I can port my load screen over from yuzu.

B3n30 commented on src/common/zstd_compression.cpp:
```diff
@@ -0,0 +1,51 @@
+namespace Common::Compression {
+
+std::vector<u8> CompressDataZSTD(const u8* source, std::size_t source_size, s32 compression_level) {
+    compression_level = std::clamp(compression_level, 1, ZSTD_maxCLevel());
```
magical number `1, is there no `ZSTD_minCLevel()`? is there no level 0?

jroweboy replied:
Funny enough there isn't in the version we used, but the recently added it upstream so i'm updating and using it here now.

B3n30 commented on src/video_core/renderer_opengl/gl_shader_disk_cache.cpp:
```diff
@@ -0,0 +1,493 @@
+
+namespace OpenGL {
+
+using ShaderCacheVersionHash = std::array<u8, 64>;
```
magic number

B3n30 commented on src/video_core/renderer_opengl/gl_shader_disk_cache.h:
```diff
@@ -0,0 +1,215 @@
+    explicit ShaderDiskCacheRaw(u64 unique_identifier, ProgramType program_type,
+                                RawShaderConfig config, ProgramCode program_code);
+    ShaderDiskCacheRaw();
+    ~ShaderDiskCacheRaw();
```
= default should be in header not in cpp. Same for other ctors/dtors in this file

B3n30 commented on src/video_core/renderer_opengl/gl_shader_disk_cache.h:
```diff
@@ -0,0 +1,215 @@
+    std::string GetBaseDir() const;
+
+    /// Get current game's title id as u64
+    u64 GetProgramID();
```
const?

jroweboy replied:
Its not actually const. it has to load the program id if it exists from AppLoader.

B3n30 commented on src/video_core/renderer_opengl/gl_shader_disk_cache.h:
```diff
@@ -0,0 +1,215 @@
+    std::string GetTitleID();
+
+    template <typename T>
+    bool SaveArrayToPrecompiled(const T* data, std::size_t length) {
```
I think this doesn't need a template and could just be `const u8* data`
Same for the function below

B3n30 commented on src/video_core/renderer_opengl/gl_shader_manager.cpp:
```diff
@@ -316,4 +431,182 @@ void ShaderProgramManager::ApplyTo(OpenGLState& state) {
+    std::vector<std::size_t> load_raws_index;
+    // Loads both decompiled and precompiled shaders from the cache. If either one is missing for
+    const auto LoadPrecompiledWorker =
+        [&](std::size_t begin, std::size_t end, const std::vector<ShaderDiskCacheRaw>& raws,
```
begin is unused

this can probably be made nicer (and maybe faster) by using iterators, same for the other lambdas below

jroweboy replied:
The reason I made them lambdas is for threaded shader loading (which was a feature that I removed). This is also why begin is unused. I don't mind changing it back, but i think its more readable this way personally so you can see which blocks are what. Just my opinion though.

B3n30 commented on src/video_core/renderer_opengl/gl_shader_manager.cpp:
```diff
@@ -316,4 +431,182 @@ void ShaderProgramManager::ApplyTo(OpenGLState& state) {
+
+    compilation_failed = false;
+
+    const auto LoadTransferable = [&](std::size_t begin, std::size_t end,
```
unused begin

jroweboy commented on src/core/hw/gpu.cpp:
```diff
@@ -11,6 +11,7 @@
 #include "common/logging/log.h"
 #include "common/microprofile.h"
 #include "common/vector_math.h"
+#include "core/core.h"
```
included for `Core::System::GetInstance()`

jroweboy commented on src/common/file_util.cpp:
```diff
@@ -546,7 +546,7 @@ std::optional<std::string> GetCurrentDir() {
     free(dir);
     return strDir;
-}
+} // namespace FileUtil
```
sorry but it really wants it https://travis-ci.org/citra-emu/citra/jobs/637758338#L576

jroweboy commented:
I've addressed all feedback and fixed the sanitize mul issue. I plan to merge this tomorrow so that we can start testing some other things. Thanks for the feedback!
